runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/retries/impl/StandardRetryStrategy.kt

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+package aws.smithy.kotlin.runtime.retries.impl
+
+import aws.smithy.kotlin.runtime.retries.*
+import kotlinx.coroutines.TimeoutCancellationException
+import kotlinx.coroutines.withTimeout
+
+/**
+ * Implements a retry strategy utilizing backoff delayer and a token bucket for rate limiting and circuit breaking. Note
+ * that the backoff delayer and token bucket work independently of each other. Either can delay retries (and the token
+ * bucket can delay the initial try). The delayer is called first so that the token bucket can refill as appropriate.
+ * @param options The options that control the functionality of this strategy.
+ * @param tokenBucket The token bucket instance. Utilizing an existing token bucket will share call capacity between
+ * those scopes.
+ * @param backoffDelayer A delayer that can back off after the initial try to spread out the retries.
+ */
+class StandardRetryStrategy(
+    val options: StandardRetryStrategyOptions,
+    private val tokenBucket: RetryTokenBucket,
+    private val backoffDelayer: BackoffDelayer,
+) : RetryStrategy {
+    /**
+     * Retry the given block of code until it's successful. Note this method throws exceptions for non-successful
+     * outcomes from retrying.
+     */
+    override suspend fun <R> retry(policy: RetryPolicy<R>, block: suspend () -> R): R =
+        withTimeout(options.maxTimeMs) {
+            doTryLoop(block, policy, 1, tokenBucket.acquireToken(), null)
+        }
+
+    /**
+     * Perform a single iteration of the try loop. Execute the block of code, evaluate the result, and take action to
+     * terminate or enter the next iteration of the try loop.
+     * @param block The block of code to retry.
+     * @param policy The [RetryPolicy] to use for evaluating the execution result.
+     * @param attempt The ordinal index of the retry loop (starting at 1).
+     * @param fromToken A [RetryToken] which grants the strategy capacity to execute a try. This token is resolved
+     * inside the function by calling [notifySuccess][RetryToken.notifySuccess],
+     * [notifyFailure][RetryToken.notifyFailure], or [scheduleRetry][RetryToken.scheduleRetry].
+     * @param previousResult The [Result] from the prior loop iteration. This is used in the case of a timeout to
+     * include in the thrown exception.
+     */
+    private tailrec suspend fun <R> doTryLoop(
+        block: suspend () -> R,
+        policy: RetryPolicy<R>,
+        attempt: Int,
+        fromToken: RetryToken,
+        previousResult: Result<R>?,
+    ): R {
+        val callResult = runCatching { block() }
+        if (callResult.exceptionOrNull() is TimeoutCancellationException) {
```

Comment:
comment

We also probably want to check for cancellation and re-throw the same exception unmodified.

```kotlin
if (callResult.exceptionOrNull() is CancellationException) { // re-throw }
```

Unless we are sure that we will end up re-throwing some exception when encountered.